File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/Body.java
package com.jcons.pdfconverter.pdfconverter;

import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;

public class Body extends List {

    private int mByteOffsetStart;
    private int mObjectNumberStart;
    private int mGeneratedObjectsCount;
    private ArrayList<IndirectObject> mObjectsList;

    public Body() {
        super();
        clear();
    }

    public int getObjectNumberStart() {
        return mObjectNumberStart;
    }

    public void setObjectNumberStart(int value) {
        mObjectNumberStart = value;
    }

    public void setByteOffsetStart(int value) {
        mByteOffsetStart = value;
    }

    public int getObjectsCount() {
        return mObjectsList.size();
    }

    private int getNextAvailableObjectNumber() {
        return ++mGeneratedObjectsCount + mObjectNumberStart;
    }

    public IndirectObject getNewIndirectObject() {
        return getNewIndirectObject(getNextAvailableObjectNumber(), 0, true);
    }

    public IndirectObject getNewIndirectObject(int number, int generation, boolean inUse) {
        IndirectObject iobj = new IndirectObject();
        iobj.setNumberID(number);
        iobj.setGeneration(generation);
        iobj.setInUse(inUse);
        return iobj;
    }

    public IndirectObject getObjectByNumberID(int number) {
        for (IndirectObject iobj : mObjectsList)
            if (iobj.getNumberID() == number)
                return iobj;

        return null;
    }

    public void includeIndirectObject(IndirectObject iobj) {
        mObjectsList.add(iobj);
    }

    private String render() {
        int offset = mByteOffsetStart;
        for (int x = 1; x <= mObjectsList.size(); x++) {
            IndirectObject iobj = getObjectByNumberID(x);
            String s = "";

            if (iobj != null){
                s = iobj.toPDFString() + "\n";
                iobj.setByteOffset(offset);
            }

            mList.add(s);
            offset += s.length();
        }
        return renderList();
    }

    @Override
    public String toPDFString() {
        return render();
    }

    public int writeTo(OutputStream stream) throws IOException {
        int offset = mByteOffsetStart;
        for (int x = 1; x <= mObjectsList.size(); x++) {
            IndirectObject iobj = getObjectByNumberID(x);
            if (iobj != null) {
                iobj.setByteOffset(offset);
                offset += iobj.writeTo(stream) + 1;
                stream.write('\n');
            }
        }
        return offset - mByteOffsetStart;
    }

    @Override
    public void clear() {
        super.clear();
        mByteOffsetStart = 0;
        mObjectNumberStart = 0;
        mGeneratedObjectsCount = 0;
        mObjectsList = new ArrayList<IndirectObject>();
    }
}

File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/Body.kt
package com.jcons.pdfconverter.pdfconverter

import java.io.IOException
import java.io.OutputStream
import java.util.*


class Body : List() {
    private var mByteOffsetStart = 0
    var objectNumberStart = 0
    private var mGeneratedObjectsCount = 0
    private var mObjectsList: ArrayList<IndirectObject>? = null

    fun setByteOffsetStart(value: Int) {
        mByteOffsetStart = value
    }

    val objectsCount: Int
        get() = mObjectsList!!.size

    private val nextAvailableObjectNumber: Int
        private get() = ++mGeneratedObjectsCount + objectNumberStart

    val newIndirectObject: IndirectObject
        get() = getNewIndirectObject(nextAvailableObjectNumber, 0, true)

    fun getNewIndirectObject(number: Int, generation: Int, inUse: Boolean): IndirectObject {
        val iobj = IndirectObject()
        iobj.numberID = number
        iobj.generation = generation
        iobj.inUse = inUse
        return iobj
    }

    fun getObjectByNumberID(number: Int): IndirectObject? {
        for (iobj in mObjectsList!!) if (iobj.numberID == number) return iobj
        return null
    }

    fun includeIndirectObject(iobj: IndirectObject) {
        mObjectsList!!.add(iobj)
    }

    private fun render(): String {
        var offset = mByteOffsetStart
        for (x in 1..mObjectsList!!.size) {
            val iobj = getObjectByNumberID(x)
            var s = ""
            if (iobj != null) {
                s = """
                    ${iobj.toPDFString()}
                    
                    """.trimIndent()
                iobj.byteOffset = offset
            }
            mList.add(s)
            offset += s.length
        }
        return renderList()
    }

    override fun toPDFString(): String {
        return render()
    }

    @Throws(IOException::class)
    fun writeTo(stream: OutputStream): Int {
        var offset = mByteOffsetStart
        for (x in 1..mObjectsList!!.size) {
            val iobj = getObjectByNumberID(x)
            if (iobj != null) {
                iobj.byteOffset = offset
                offset += iobj.writeTo(stream) + 1
                stream.write('\n'.toInt())
            }
        }
        return offset - mByteOffsetStart
    }

    override fun clear() {
        super.clear()
        mByteOffsetStart = 0
        objectNumberStart = 0
        mGeneratedObjectsCount = 0
        mObjectsList = ArrayList()
    }

    init {
        clear()
    }
}
